Describe query AST

Totktonada opened the pull request:
Added placeholders for \<inputObject\> and \<variable\> nodes to place
corresponding descriptions. We'll hopefully implement strictier
validation and will process those nodes anyway.

Totktonada committed:
Describe query AST

Totktonada commented:
Source: https://github.com/tarantool/graphql.0/commit/87bb0c8d602c88b2e38b316c95a7dba78b1ec6ae

RunsFor commented on graphql/validate.lua:
```diff
@@ -35,6 +46,51 @@ local visitors = {
+  --     kind = 'name',
+  --     value = <string>,
+  --   } (optional),
+  --   selectionSet = <selectionSet>,
```
according to [Oct. 2016 spec](http://spec.graphql.org/October2016/#SelectionSet) it should be `list of <Selection>`

Totktonada replied:
Here we describe the AST produced by the [graphql query parser](https://github.com/tarantool/luagraphqlparser). It creates a node for the `selectionSet`, which contains the list of `selection` nodes. There is no contradiction with the syntactic grammar provided by the specification.

RunsFor commented on graphql/validate.lua:
```diff
@@ -35,6 +46,51 @@ local visitors = {
+  -- <listType>, <nonNullType>
+  --
+  -- {
+  --   kind = one of ('listType', 'nonNullType'),
```
According to this definition, `nonNullType` may be `nonNullType` (which is recursive), instead it should be one of `namedType` or `listType`

Totktonada replied:
It seems correct. Fixed.

RunsFor commented on graphql/validate.lua:
```diff
@@ -73,6 +141,22 @@ local visitors = {
+  --   selectionSet = <selectionSet> (optional),
+  --   alias = <alias> (optional),
+  --   arguments = list of <argument> (optional),
+  --   directives = list of <directive> (optional),
```
I suppose, required `name` is missing here

Totktonada replied:
Looks correct. Fixed.

RunsFor commented on graphql/validate.lua:
```diff
@@ -218,6 +320,17 @@ local visitors = {
+  --     value = <string>,
+  --   },
+  --   typeCondition = <namedType>,
+  --   selectionSet = <selectionSet>,
```
I suppose, [optional directives](http://spec.graphql.org/October2016/#FragmentDefinition) are missing

Totktonada replied:
It was not there, because the old lpeg based parser does not support it. I looked into [luagraphqlparser](https://github.com/tarantool/luagraphqlparser) and it [appears](https://github.com/tarantool/luagraphqlparser/issues/2) that it does not support it too. We fixed this with @olegrok and it'll be supported soon.

So, I added the directives field here.

RunsFor commented on graphql/validate.lua:
```diff
@@ -245,6 +358,41 @@ local visitors = {
+  -- * <string>
+  -- * <boolean>
+  -- * <float>
+  -- * <int>
```
suppose, [null value](http://spec.graphql.org/October2016/#sec-Null-Value) is missing here

Totktonada replied:
It is not supported in the old lpeg based parser, which was used at times, when I wrote this patch. Now [luagraphqlparser](https://github.com/tarantool/luagraphqlparser) is there and it is supported.

So, I added it.

RunsFor commented on graphql/validate.lua:
```diff
@@ -262,6 +410,44 @@ local visitors = {
+  --   },
+  -- }
+  --[[
+  inputObject = {
```
Probably full featured example would be more helpful here

Totktonada replied:
I would stick with just formal description for now.

RunsFor commented on graphql/validate.lua:
```diff
@@ -262,6 +410,44 @@ local visitors = {
+  --   },
+  -- }
+  --[[
+  variable = {
```
Here as well

Totktonada replied:
In the AST we have only variable name here. It looks pretty simple and, to be honest, I don't know, which kind of example would add something there. If you have an idea, please, let me know.

RunsFor reviewed (changes requested):
Thanks for your PR. Sorry for delay with reviewing your changes.

I think we need to determine which version of specification we are supporting right now, since latest revisions went far away already.

In review I relied on [this version](http://spec.graphql.org/October2016) of spec. Correct me if I chose the wrong one.

Totktonada commented:
@RunsFor Hi! Please, excuse me for the delay too. And thank you for the in-depth review!

I rebased the patch at top of current master and fixed all comments except ones about providing examples. I'll return there when time permits. I'll mark the PR as draft until I'll resolve all comments.

Totktonada commented on graphql/validate.lua:
```diff
@@ -270,6 +431,15 @@ local visitors = {
+  --     name = <string>,
+  --     value = <value>,
+  --   },
+  -- }
```
An input object may contain a another input object as a field. It is not reflected there.

Totktonada replied:
From above:

```lua
  -- <value> is one of the following:
  --  
  -- * <variable>
  -- * <inputObject>
  -- * <list>
  -- * <enum>
  -- * <string>
  -- * <boolean>
  -- * <float>
  -- * <int>
  -- * <null>
```

So an inputObject actually can contain an inputObject as a field according to the AST description. Everyting looks okay here.

Totktonada committed:
doc: generate query AST graph

Totktonada commented:
Updated with two changes.

* Resolved ambiguity with `<string>`.

  The `<string>` in my GraphQL query representation now means two different objects: an AST node for a value of the string type (it is written in double quotes in a query) and a Lua string. Replaced Lua strings with `'...'`.

  Also removed `it is 'null' for <null>` comment, because here everything the same for all values type: the 'value' field contains a Lua string as it is wriiten in the query.

  Squashed into the original commit.

* Added generation of a picture with query AST.

  Just tried it for fun. Don't know, whether it'll be useful.

  Added as a separate commit.

no1seman commented:
@Totktonada Is this the same: https://github.com/tarantool/graphqlapi/blob/master/docs/types.md?

Totktonada commented:
It is query AST.

Totktonada commented:
@RunsFor Are you interested to look into the pull request again?

Totktonada commented:
@olegrok What do you think on this? I would like to proceed in either way: drop at all, push first patch or push both patches.

olegrok commented:
> @olegrok What do you think on this? I would like to proceed in either way: drop at all, push first patch or push both patches.

I like both patches. However currently I think it's better to place them to luagraphqlparser repo

Totktonada commented:
> I like both patches. However currently I think it's better to place them to luagraphqlparser repo

Seems meaningful. I'll close this pull request and will open another one in graphqlparser if time will permit.
